File: src/main/java/com/amberclient/modules/render/xray/Xray.java
package com.amberclient.modules.render.xray;

import com.amberclient.utils.module.Module;
import com.amberclient.utils.module.ConfigurableModule;
import com.amberclient.utils.module.ModuleSettings;
import net.fabricmc.fabric.api.client.rendering.v1.WorldRenderEvents;
import net.minecraft.client.MinecraftClient;
import net.minecraft.text.Text;
import net.minecraft.util.math.ChunkPos;
import org.apache.logging.log4j.Logger;
import org.apache.logging.log4j.LogManager;

import java.util.List;
import java.util.ArrayList;

public class Xray extends Module implements ConfigurableModule {
    public static final String MOD_ID = "amberclient-xray";
    public static final Logger LOGGER = LogManager.getLogger(MOD_ID);
    private ChunkPos lastPlayerChunk;
    private boolean wasActive = false;

    private final ModuleSettings chunkRadius;
    private final ModuleSettings exposedOnly;
    private final ModuleSettings showLava;
    private final ModuleSettings oreSim;
    private final List<ModuleSettings> settings;

    public Xray() {
        super("XRay", "Shows the outlines of the selected ores in the chunk", "Render");

        // Settings
        chunkRadius = new ModuleSettings("Chunk Radius", "Number of chunks to scan around the player", 1.0, 1.0, 8.0, 1.0);
        exposedOnly = new ModuleSettings("Exposed Only", "Show only ores exposed to air", false);
        showLava = new ModuleSettings("Show Lava", "Show lava with the ores", false);
        oreSim = new ModuleSettings("OreSim", "Use world seed to predict ore generation instead of scanning existing blocks", false);

        settings = new ArrayList<>();
        settings.add(chunkRadius);
        settings.add(exposedOnly);
        settings.add(showLava);
        settings.add(oreSim);

        WorldRenderEvents.AFTER_TRANSLUCENT.register(RenderOutlines::render);
    }

    @Override
    public void onEnable() {
        SettingsStore.getInstance().get().setActive(true);
        SettingsStore.getInstance().get().setExposedOnly(exposedOnly.getBooleanValue());
        SettingsStore.getInstance().get().setShowLava(showLava.getBooleanValue());
        SettingsStore.getInstance().get().setHalfRange((int) chunkRadius.getDoubleValue());
        SettingsStore.getInstance().get().setOreSim(oreSim.getBooleanValue());

        ScanTask.resetLocationTracking();

        MinecraftClient client = MinecraftClient.getInstance();
        if (client.player != null) {
            ScanTask.runTask(client.player.getChunkPos(), SettingsStore.getInstance().get().getHalfRange(), true);
            lastPlayerChunk = null;
        }
        wasActive = true;
    }

    @Override
    public void onDisable() {
        SettingsStore.getInstance().get().setActive(false);
        ScanTask.renderQueue.clear();
        lastPlayerChunk = null;
        wasActive = false;

        OreSim.chunkRenderers.clear();
        OreSim.oreConfig = null;
    }

    @Override
    public void onTick() {
        if (SettingsStore.getInstance().get().isActive()) {
            MinecraftClient client = MinecraftClient.getInstance();
            if (client.player == null) return;

            ChunkPos currentChunk = client.player.getChunkPos();
            if (!currentChunk.equals(lastPlayerChunk) || !wasActive) {
                ScanTask.runTask(currentChunk, SettingsStore.getInstance().get().getHalfRange());
                lastPlayerChunk = currentChunk;
                wasActive = true;
            }
        }
    }

    @Override
    public List<ModuleSettings> getSettings() {
        return settings;
    }

    @Override
    public void onSettingChanged(ModuleSettings setting) {
        MinecraftClient client = MinecraftClient.getInstance();
        if (client.player == null) return;

        if (setting == exposedOnly) {
            SettingsStore.getInstance().get().setExposedOnly(exposedOnly.getBooleanValue());
            client.player.sendMessage(
                    Text.literal("§6Exposed Only: §l" + (exposedOnly.getBooleanValue() ? "ON" : "OFF")),
                    true
            );
            if (SettingsStore.getInstance().get().isActive()) {
                ScanTask.runTask(client.player.getChunkPos(), SettingsStore.getInstance().get().getHalfRange(), true);
            }
        } else if (setting == chunkRadius) {
            SettingsStore.getInstance().get().setHalfRange((int) chunkRadius.getDoubleValue());
            client.player.sendMessage(
                    Text.literal("§6Chunk Range: §l" + (int) chunkRadius.getDoubleValue() + " chunks"),
                    true
            );
            if (SettingsStore.getInstance().get().isActive()) {
                ScanTask.runTask(client.player.getChunkPos(), SettingsStore.getInstance().get().getHalfRange(), true);
            }
        } else if (setting == showLava) {
            SettingsStore.getInstance().get().setShowLava(showLava.getBooleanValue());
            client.player.sendMessage(
                    Text.literal("§6Show Lava: §l" + (showLava.getBooleanValue() ? "ON" : "OFF")),
                    true
            );
            if (SettingsStore.getInstance().get().isActive()) {
                ScanTask.runTask(client.player.getChunkPos(), SettingsStore.getInstance().get().getHalfRange(), true);
            }
        } else if (setting == oreSim) { // Nouveau paramètre
            SettingsStore.getInstance().get().setOreSim(oreSim.getBooleanValue());
            client.player.sendMessage(
                    Text.literal("§6OreSim: §l" + (oreSim.getBooleanValue() ? "ON" : "OFF")),
                    true
            );
            if (SettingsStore.getInstance().get().isActive()) {
                ScanTask.runTask(client.player.getChunkPos(), SettingsStore.getInstance().get().getHalfRange(), true);
            }
        }
    }
}

File: src/main/java/com/amberclient/modules/render/xray/Xray.kt
package com.amberclient.modules.render.xray

import com.amberclient.utils.module.Module
import com.amberclient.utils.module.ConfigurableModule
import com.amberclient.utils.module.ModuleSettings
import net.fabricmc.fabric.api.client.rendering.v1.WorldRenderEvents
import net.minecraft.client.MinecraftClient
import net.minecraft.text.Text
import net.minecraft.util.math.ChunkPos
import org.apache.logging.log4j.Logger
import org.apache.logging.log4j.LogManager

class Xray : Module("XRay", "Shows the outlines of the selected ores in the chunk", "Render"), ConfigurableModule {

    companion object {
        const val MOD_ID = "amberclient-xray"
        val LOGGER: Logger = LogManager.getLogger(MOD_ID)
    }

    private var lastPlayerChunk: ChunkPos? = null
    private var wasActive = false

    private val chunkRadius: ModuleSettings
    private val exposedOnly: ModuleSettings
    private val showLava: ModuleSettings
    private val settings: List<ModuleSettings>

    init {
        // Settings
        chunkRadius = ModuleSettings("Chunk Radius", "Number of chunks to scan around the player", 1.0, 1.0, 8.0, 1.0)
        exposedOnly = ModuleSettings("Exposed Only", "Show only ores exposed to air", false)
        showLava = ModuleSettings("Show Lava", "Show lava with the ores", false)

        settings = listOf(chunkRadius, exposedOnly, showLava)

        WorldRenderEvents.AFTER_TRANSLUCENT.register(RenderOutlines::render)
    }

    override fun onEnable() {
        val settingsStore = SettingsStore.getInstance().get()
        settingsStore.isActive = true
        settingsStore.isExposedOnly = exposedOnly.booleanValue
        settingsStore.isShowLava = showLava.booleanValue
        settingsStore.halfRange = chunkRadius.doubleValue.toInt()

        ScanTask.resetLocationTracking()

        val client = MinecraftClient.getInstance()
        client.player?.let { player ->
            ScanTask.runTask(player.chunkPos, settingsStore.halfRange, true)
            lastPlayerChunk = null
        }
        wasActive = true
    }

    override fun onDisable() {
        SettingsStore.getInstance().get().isActive = false
        ScanTask.renderQueue.clear()
        lastPlayerChunk = null
        wasActive = false
    }

    override fun onTick() {
        val settingsStore = SettingsStore.getInstance().get()
        if (settingsStore.isActive) {
            val client = MinecraftClient.getInstance()
            val player = client.player ?: return

            val currentChunk = player.chunkPos
            if (currentChunk != lastPlayerChunk || !wasActive) {
                ScanTask.runTask(currentChunk, settingsStore.halfRange)
                lastPlayerChunk = currentChunk
                wasActive = true
            }
        }
    }

    override fun getSettings(): List<ModuleSettings> = settings

    override fun onSettingChanged(setting: ModuleSettings) {
        val client = MinecraftClient.getInstance()
        val player = client.player ?: return
        val settingsStore = SettingsStore.getInstance().get()

        when (setting) {
            exposedOnly -> {
                settingsStore.isExposedOnly = exposedOnly.booleanValue
                player.sendMessage(
                    Text.literal("§6Exposed Only: §l${if (exposedOnly.booleanValue) "ON" else "OFF"}"),
                    true
                )
                if (settingsStore.isActive) {
                    ScanTask.runTask(player.chunkPos, settingsStore.halfRange, true)
                }
            }
            chunkRadius -> {
                settingsStore.halfRange = chunkRadius.doubleValue.toInt()
                player.sendMessage(
                    Text.literal("§6Chunk Range: §l${chunkRadius.doubleValue.toInt()} chunks"),
                    true
                )
                if (settingsStore.isActive) {
                    ScanTask.runTask(player.chunkPos, settingsStore.halfRange, true)
                }
            }
            showLava -> {
                settingsStore.isShowLava = showLava.booleanValue
                player.sendMessage(
                    Text.literal("§6Show Lava: §l${if (showLava.booleanValue) "ON" else "OFF"}"),
                    true
                )
                if (settingsStore.isActive) {
                    ScanTask.runTask(player.chunkPos, settingsStore.halfRange, true)
                }
            }
        }
    }
}